### PR TITLE
Working Flysky Hall Gimbal implementation

### PR DESCRIPTION
**Pull request by rafolg**
This is finally working and is thanks to the fantastic work of Janek by enabling the idle callback in UART4. The gimbal works with a minimum buffer size of 128 bytes, any lower than that and the radio reboots constantly

### CODE REVIEW

File: radio/src/targets/flysky/board.cpp [hunk 1]
```diff
@@ -224,6 +224,7 @@ void boardInit()
   init2MhzTimer();
   init5msTimer();
   __enable_irq();
+  aux2SerialSetIdleCb(flysky_gimbal_loop);
```
Comment (ajjjjjjjj) [inline radio/src/targets/flysky/board.cpp:227]: Please call flysky_gimbal_init here and register callback from it. If init implementation is causing trouble please comment it out but leave callback registration there.

File: radio/src/targets/flysky/CMakeLists.txt [hunk 2]
```diff
@@ -136,6 +136,7 @@ set(TARGET_SRC
   i2c_driver.cpp
   crc_driver.cpp
   ../common/arm/stm32/mixer_scheduler_driver.cpp
+  ../common/arm/stm32/flysky_gimbal_driver.cpp
```
Comment (ajjjjjjjj) [inline radio/src/targets/flysky/CMakeLists.txt:139]: Guard using FLYSKY_GIMBAL

File: radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+      HallProtocol.msg_OK = 0;
+      HallProtocol.stickState = HallProtocol.data[HallProtocol.length - 1];
+
+      switch (HallProtocol.hallID.hall_Id.receiverID)
+      {
+        case TRANSFER_DIR_TXMCU:
+          if (HallProtocol.hallID.hall_Id.packetID == FLYSKY_HALL_RESP_TYPE_VALUES) {
+            int16_t* p_values = (int16_t*)HallProtocol.data;
+            for (uint8_t i = 0; i < 4; i++) {
+              hall_adc_values[i] = FLYSKY_OFFSET_VALUE - (p_values[i] >> 1);
```
Comment (ajjjjjjjj) [inline radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp:170]: Please write directly to adcValues and remove changes in getADC(). I know it is causing conflict with existing gimballs sampling so in adcInit() remove sampling of standard gimballs related channels and writing to it when FLYSKY_GIMBAL is enabled.

File: radio/src/opentx.h [hunk 4]
```diff
@@ -580,6 +580,7 @@ PACK(struct GlobalData {
   uint8_t unexpectedShutdown;
   uint8_t usbConnect;
   uint8_t cToolRunning;
+uint8_t flyskygimbals;
```
Comment (ajjjjjjjj) [inline radio/src/opentx.h:583]: After suggested changes this can go back to flysky gimball driver space and keep the same name as in source (EdgeTX)

**Review by ajjjjjjjj — CHANGES_REQUESTED** [review]
Great work. 
I will take care of CRC16 calculation moving to hardware peripheral.

File: radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp (hunk 3 above)
Comment (rafolg) [reply radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp:170]: Yeah, I tried that but by disabling adc_channel_0, 1, 2, and 3, stuffs up the order of the switches, knobs and sticks. The best way would be to mask channels 0-3 like edgeTX does but could not find the similar adc code here

File: radio/src/targets/flysky/CMakeLists.txt (hunk 2 above)
Comment (rafolg) [reply radio/src/targets/flysky/CMakeLists.txt:139]: Done

File: radio/src/targets/flysky/board.cpp (hunk 1 above)
Comment (rafolg) [reply radio/src/targets/flysky/board.cpp:227]: flysky_gimbal_init can not be called here as it does not allow the gimbal to communicate. I have comment it out and just left idlecallback here. Works well this way

File: radio/src/targets/flysky/CMakeLists.txt (hunk 2 above)
Comment (rafolg) [reply radio/src/targets/flysky/CMakeLists.txt]: Done

File: radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp (hunk 3 above)
Comment (ajjjjjjjj) [reply radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp:173]: Adjusting FIRST_ANALOG_ADC and NUM_ANALOGS_ADC does not help?

Comment (rafolg) [reply radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp:175]: No dice, even adjusting those it stuffs up the order of everything else.

Comment (rafolg) [reply radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp:175]: Ohh, actually, I havent changed the DMA_MemoryBaseAddr

File: radio/src/opentx.h (hunk 4 above)
Comment (rafolg) [reply radio/src/opentx.h:583]: Fixed!

File: radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp (hunk 3 above)
Comment (rafolg) [reply radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp]: It is all now fixed

**Comment by rafolg** [comment]
4/4 requests completed

File: radio/src/targets/flysky/board.cpp (hunk 1 above)
Comment (ajjjjjjjj) [reply radio/src/targets/flysky/board.cpp:227]: It should be called here to register serial callback (see original implementation), other part of implementation can be commented out with comment why.

File: radio/src/targets/flysky/crc_driver.cpp [hunk 5]
```diff
@@ -13,7 +13,9 @@ void crcInit() {
 }
 
 uint8_t crc8_hw(const uint8_t * ptr, uint32_t len) {
+  CRC->INIT = CRC8_INIT_VAL;
```
Comment (ajjjjjjjj) [inline radio/src/targets/flysky/crc_driver.cpp:16]: After this change crcInit is not needed and must be removed.

File: radio/src/targets/flysky/adc_driver.cpp [hunk 6]
```diff
@@ -137,7 +156,11 @@ void adcInit()
   dma_init.DMA_BufferSize = NUM_ANALOGS;
   // source and destination start addresses
   dma_init.DMA_PeripheralBaseAddr = (uint32_t)&ADC1->DR;
+  #if defined(FLYSKY_GIMBAL)
+  dma_init.DMA_MemoryBaseAddr = (uint32_t)adcMemoryValueAddr;
```
Comment (ajjjjjjjj) [inline radio/src/targets/flysky/adc_driver.cpp:160]: Reference &adcValues[FIRST_ANALOG_ADC] directly and remove define condition

File: radio/src/targets/flysky/adc_driver.cpp [hunk 7]
```diff
@@ -100,9 +114,14 @@ void adcInit()
   ADC_Init(ADC1, &adc_init);
 
   // configure each channel
+  #if defined(FLYSKY_GIMBAL)
```
Comment (ajjjjjjjj) [inline radio/src/targets/flysky/adc_driver.cpp:117]: Use !defined and keep this line same for both conditions to simplify it

File: radio/src/targets/flysky/crc_driver.cpp (hunk 5 above)
Comment (rafolg) [reply radio/src/targets/flysky/crc_driver.cpp:16]: Even if crc16 doesn't use the same init value? CRC uses 0xFFFF.

Don't we need to change it back every time the code runs either crc8 or crc16?

Comment (ajjjjjjjj) [reply radio/src/targets/flysky/crc_driver.cpp:16]: It is now set up in every crc function so initCrc does not make sense

File: radio/src/targets/flysky/adc_driver.cpp (hunk 6 above)
Comment (rafolg) [reply radio/src/targets/flysky/adc_driver.cpp:160]: Thanks, fixed now

File: radio/src/targets/flysky/adc_driver.cpp (hunk 7 above)
Comment (rafolg) [reply radio/src/targets/flysky/adc_driver.cpp:117]: Thanks, done!

File: radio/src/targets/flysky/crc_driver.cpp (hunk 5 above)
Comment (rafolg) [reply radio/src/targets/flysky/crc_driver.cpp:16]: Ohh I get it, sorry I originally misread your comment half way through the night. All fixed now

Comment (ajjjjjjjj) [reply radio/src/targets/flysky/crc_driver.cpp:16]: Remove it completely instead of commenting out, not needed.

**Review by ajjjjjjjj — APPROVED** [review]

**Pull request merged**